app/browser/api/ledger.js

```diff
@@ -0,0 +1,2187 @@
+      const paymentId = (paymentIdLine.match(paymentIdPattern) || [])[1]
+
+      const passphrasePattern = new RegExp([locale.translation('ledgerBackupText4'), '(.+)$'].join(' '))
+      const passphrase = (passphraseLine.match(passphrasePattern) || [])[1]
+
+      keys = {
+        paymentId,
+        passphrase
+      }
+    } catch (exc) {
+      state = logError(state, exc, 'recoveryWallet')
+    }
+  }
+
+  return {
+    state,
+    keys
+  }
+}
+
+const getPublisherData = (result, scorekeeper) => {
+  let duration = result.duration
+
+  let data = {
+    verified: result.options.verified || false,
+    site: result.publisherKey,
+    views: result.visits,
+    duration: duration,
+    daysSpent: 0,
+    hoursSpent: 0,
+    minutesSpent: 0,
+    secondsSpent: 0,
+    faviconURL: result.faviconURL,
+    score: result.scores[scorekeeper],
+    pinPercentage: result.pinPercentage,
+    weight: result.pinPercentage
+  }
+  // HACK: Protocol is sometimes blank here, so default to http:// so we can
+  // still generate publisherURL.
+  data.publisherURL = (result.protocol || 'http:') + '//' + result.publisherKey
+
+  if (duration >= miliseconds.day) {
+    data.daysSpent = Math.max(Math.round(duration / miliseconds.day), 1)
+  } else if (duration >= miliseconds.hour) {
+    data.hoursSpent = Math.max(Math.floor(duration / miliseconds.hour), 1)
+    data.minutesSpent = Math.round((duration % miliseconds.hour) / miliseconds.minute)
+  } else if (duration >= miliseconds.minute) {
+    data.minutesSpent = Math.max(Math.round(duration / miliseconds.minute), 1)
+    data.secondsSpent = Math.round((duration % miliseconds.minute) / miliseconds.second)
+  } else {
+    data.secondsSpent = Math.max(Math.round(duration / miliseconds.second), 1)
+  }
+
+  return data
+}
+
+const normalizePinned = (dataPinned, total, target, setOne) => dataPinned.map((publisher) => {
+  let newPer
+  let floatNumber
+
+  if (setOne) {
+    newPer = 1
+    floatNumber = 1
+  } else {
+    floatNumber = (publisher.pinPercentage / total) * target
+    newPer = Math.floor(floatNumber)
+    if (newPer < 1) {
+      newPer = 1
+    }
+  }
+
+  publisher.weight = floatNumber
+  publisher.pinPercentage = newPer
+  return publisher
+})
+
+// courtesy of https://stackoverflow.com/questions/13483430/how-to-make-rounded-percentages-add-up-to-100#13485888
+const roundToTarget = (l, target, property) => {
+  let off = target - underscore.reduce(l, (acc, x) => { return acc + Math.round(x[property]) }, 0)
+
+  return underscore.sortBy(l, (x) => Math.round(x[property]) - x[property])
+    .map((x, i) => {
+      x[property] = Math.round(x[property]) + (off > i) - (i >= (l.length + off))
+      return x
+    })
+}
+
+// TODO we should convert this function and all related ones into immutable
+// TODO merge publishers and publisherData that is created in getPublisherData
+// so that we don't need to create new Map every single time
+const synopsisNormalizer = (state, changedPublisher) => {
```

Comment:
This function is HUGE and should be broken out. However, that's outside the scope of this PR 😄 

This could use a few good unit tests though- ex: handling when pinned total total > 100%, less than 100%, etc for publishers

Reply:
agree will try to move it in the next PR's